File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums)-1

        while l <= r:
            m = (l+r) // 2
            print(m)
            if nums[m] == target:
                return m

            if nums[l] <= nums[m]:
                # M on Left Side
                if nums[l] <= target < nums[m]:
                    r = m-1
                else:
                    l = m+1
            else:
                # M on Right Side
                if nums[m] < target <= nums[r]:
                    l = m+1
                else:
                    r = m-1

        return -1

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-1.swift
class Solution {
    func findMin(_ nums: [Int]) -> Int {
        var r = nums.count - 1

        var ret = nums[0]
        var l = 0
        
        while l <= r {
            // Bounds are in the right half
            if nums[l] < nums[r]{
                ret = min(ret, nums[l])
                break
            }

            let m = (l+r) / 2
            ret = min(ret, nums[m])
            if nums[m] >=  nums[l] {
                // Search Right
                l = m+1
            } else {
                // Search Left
                r = m-1
            }
        }
        return ret
    }
}
